helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -329,32 +330,34 @@ public void close() {
     disconnect();
   }
 
-  private void updateGCTimer(String rootPath, String currentVersion) {
-    TimerTask gcTask = new TimerTask() {
-      @Override
-      public void run() {
+  private synchronized void updateGCTimer(String rootPath, long currentVersion) {
+    if (_gcTaskFuture != null) {
+      _gcTaskFuture.cancel(false);
+    }
```

Comment:
Do we really need to cancel? What if that causes incomplete deletion of stale versions? I think we should let it queue up?

Reply:
This cancel will only cancel the pending tasks. For the pending tasks that are queued up, they will only cause extra children list read. The later operation will be the same. And if there is risk of incomplete deleting, then that concern lives no matter we let it queued up or not. That issue (if exists) should be addressed separately.